kbfsedits/tlf_history.go

```diff
@@ -0,0 +1,354 @@
+				fileEvent{newName: eventFilename}
+		}
+
+		// The renamed file overwrote any existing file with the new
+		// name.
+		r.fileEvents[eventFilename] = fileEvent{delete: true}
+	case NotificationDelete:
+		r.fileEvents[eventFilename] = fileEvent{delete: true}
+	}
+	return false
+}
+
+func (th *TlfHistory) recomputeLocked() (
+	history writersByRevision, writersWhoNeedMore map[string]bool) {
+	writersWhoNeedMore = make(map[string]bool)
+
+	// Copy the writer notifications into a heap.
+	var writersHeap writersByRevision
+	for _, wn := range th.byWriter {
+		wnCopy := writerNotifications{
+			writerName:    wn.writerName,
+			notifications: make(notificationsByRevision, len(wn.notifications)),
+		}
+		copy(wnCopy.notifications, wn.notifications)
+		writersHeap = append(writersHeap, &wnCopy)
+	}
+	heap.Init(&writersHeap)
+
+	r := newRecomputer()
+
+	// Iterate through the heap.  The writer with the next highest
+	// revision will always be at index 0.  Process that writer's
+	// first notification, then remove it and fix the heap so that the
+	// next highest revision is at index 0.  That way events that
+	// happen more recently (like deletes and renames) can be taken
+	// into account when looking at older events.
+	for writersHeap.Len() > 0 {
+		nextWriter := writersHeap[0].writerName
+		nextNotification := writersHeap[0].notifications[0]
+		doTrim := r.processNotification(nextWriter, nextNotification)
+
+		// Remove that notification, and fix the heap because this
+		// writer has a different newest revision.
+		if doTrim {
+			// Trim all earlier revisions because they won't be needed
+			// for the cached history.
+			numProcessed := r.numProcessed[nextWriter]
+			th.byWriter[nextWriter].notifications =
+				th.byWriter[nextWriter].notifications[numProcessed:]
+		} else {
+			writersHeap[0].notifications = writersHeap[0].notifications[1:]
+		}
+		if len(writersHeap[0].notifications) == 0 || doTrim {
+			heap.Pop(&writersHeap)
+		} else {
+			heap.Fix(&writersHeap, 0)
+		}
+	}
+
+	history = make(writersByRevision, 0, len(r.byWriter))
+	for writerName := range th.byWriter {
+		wn := r.byWriter[writerName]
+		if wn != nil && len(wn.notifications) > 0 {
+			history = append(history, wn)
+		}
+		if wn == nil || len(wn.notifications) < maxEditsPerWriter {
+			writersWhoNeedMore[writerName] = true
+		}
+	}
+	sort.Sort(history)
+	// Garbage-collect any writers that don't appear in the history.
+	for i := maxWritersPerHistory; i < len(history); i++ {
+		delete(th.byWriter, history[i].writerName)
+		delete(writersWhoNeedMore, history[i].writerName)
+	}
+	if len(history) > maxWritersPerHistory {
+		history = history[:maxWritersPerHistory]
+	}
+	th.computed = true
+	th.cachedHistory = history
+	return history, writersWhoNeedMore
+}
+
+func (th *TlfHistory) getHistoryIfCached() (
+	cached bool, history writersByRevision) {
+	th.lock.RLock()
+	defer th.lock.RUnlock()
+	return th.computed, th.cachedHistory
```

Comment:
Should this return `th.computed, nil` if `th.computed == false` because of the method name? Not a big deal though since it's only used in `getHistory()` below.